### PR TITLE
*: fix possible race in stats update

### PR DESCRIPTION
**Pull request by lamxTyler**
It is dangerous to update the stats simultaneously.
When the stats lease is 0,  it means we are in test mode, so it is safe to update it directly.
When the stats lease is greater than 0, all the updates must go through the `UpdateTableStatsLoop`.
PTAL @hanfei1991 @winoros

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -78,40 +78,48 @@ func (e *AnalyzeExec) Next() (*Row, error) {
...
 	}
 	close(taskCh)
-	results := make([]analyzeResult, 0, len(e.tasks))
+	dom := sessionctx.GetDomain(e.ctx)
+	lease := dom.StatsHandle().Lease
+	if lease > 0 {
+		for i := 0; i < len(e.tasks); i++ {
+			result := <-resultCh
+			if result.Err != nil {
+				return nil, errors.Trace(err)
```
Comment (hanfei1991) [inline executor/analyze.go]: If some err is nil, we can log it and continue. It's ok to return any error in line 102

File: statistics/handle.go [hunk 2]
```diff
@@ -60,16 +61,32 @@ func (h *Handle) Clear() {
...
+		listHead:        &SessionStatsCollector{mapper: make(tableDeltaMap)},
+		globalMap:       make(tableDeltaMap),
+		Lease:           lease,
 	}
 	handle.statsCache.Store(statsCache{})
 	return handle
 }
 
+// AnalyzeResult is used to represent analyze result.
+type AnalyzeResult struct {
```
Comment (hanfei1991) [inline statistics/handle.go]: AnalyzeResult is about building, which can be put in builder.go

**Comment by hanfei1991** [comment]
The rest is ok

File: domain/domain.go [hunk 3]
```diff
@@ -526,18 +526,25 @@ func (do *Domain) UpdateTableStatsLoop(ctx context.Context) error {
...
 				return
-			// This channel is sent only by ddl owner. Only owner know if this ddl is done or not.
+			// This channel is sent only by ddl owner or the drop stats executor.
 			case t := <-do.statsHandle.DDLEventCh():
-				err := do.statsHandle.HandleDDLEvent(t)
+				err = do.statsHandle.HandleDDLEvent(t)
 				if err != nil {
 					log.Error(errors.ErrorStack(err))
 				}
+			case t := <-do.statsHandle.AnalyzeResultCh():
```
Comment (tiancaiamao) [inline domain/domain.go]: Will `AnalyzeResultCh` be closed and `t` get a uninitialized value?

File: executor/analyze.go [hunk 4]
```diff
@@ -78,40 +78,48 @@ func (e *AnalyzeExec) Next() (*Row, error) {
...
 	}
 	for _, task := range e.tasks {
 		taskCh <- task
 	}
 	close(taskCh)
-	results := make([]analyzeResult, 0, len(e.tasks))
+	dom := sessionctx.GetDomain(e.ctx)
+	lease := dom.StatsHandle().Lease
+	if lease > 0 {
+		for i := 0; i < len(e.tasks); i++ {
```
Comment (tiancaiamao) [inline executor/analyze.go]: How about
```
for result, ok := range resultCh {
```

File: executor/analyze.go [hunk 5]
```diff
@@ -78,40 +78,48 @@ func (e *AnalyzeExec) Next() (*Row, error) {
...
+			result := <-resultCh
+			if result.Err != nil {
+				return nil, errors.Trace(err)
+			}
+			result.Ctx = e.ctx
+			dom.StatsHandle().AnalyzeResultCh() <- &result
```
Comment (tiancaiamao) [inline executor/analyze.go]: Will this line block?

File: executor/analyze.go [hunk 6]
```diff
@@ -78,40 +78,48 @@ func (e *AnalyzeExec) Next() (*Row, error) {
...
-	lease := dom.StatsHandle().Lease
-	if lease > 0 {
-		// We sleep two lease to make sure other tidb node has updated this node.
-		time.Sleep(lease * 2)
-	} else {
-		err := dom.StatsHandle().Update(GetInfoSchema(e.ctx))
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+	err = dom.StatsHandle().Update(GetInfoSchema(e.ctx))
```
Comment (tiancaiamao) [inline executor/analyze.go]: dom.StatsHandle().Update is called in different goroutines, is this function thread safe?

Comment (lamxTyler) [inline executor/analyze.go]: Only when stats lease equal 0, it is called in different places, but in this case,  it  is normally in  test mode.

File: domain/domain.go (hunk 3 above)
Comment (lamxTyler) [inline domain/domain.go]: It seems that it will not be closed?

File: executor/analyze.go (hunk 5 above)
Comment (lamxTyler) [inline executor/analyze.go]: You mean when the channel is full?

**Comment by tiancaiamao** [comment]
LGTM

**Comment by lamxTyler** [comment]
PTAL @hanfei1991 @winoros

File: statistics/handle.go [hunk 7]
```diff
@@ -39,9 +39,10 @@ type Handle struct {
 	// PrevLastVersion will be assigned by LastVersion every time Update is called.
 	PrevLastVersion uint64
 	statsCache      atomic.Value
-	// ddlEventCh is a channel to notify a ddl operation has happened. It is sent only by owner and read by stats handle.
-	ddlEventCh chan *ddl.Event
-
+	// ddlEventCh is a channel to notify a ddl operation has happened.
+	// It is sent only by owner or the drop stats executor, and read by stats handle.
+	ddlEventCh      chan *ddl.Event
+	analyzeResultCh chan *AnalyzeResult
```
Comment (hanfei1991) [inline statistics/handle.go]: add some simple comments to explain why we need this channel to sync results.

**Pull request merged**